test-app/tests/integration/components/headless-form-native-validation-test.gts

```diff
@@ -0,0 +1,754 @@
+      const data = { firstName: 'foo123', lastName: 'Smith' };
+      const formValidateCallback = ({ firstName }: { firstName: string }) =>
+        firstName.charAt(0).toUpperCase() !== firstName.charAt(0)
+          ? {
+              firstName: [
+                {
+                  type: 'uppercase',
+                  value: firstName,
+                  message: 'First name must be upper case!',
+                },
+              ],
+            }
+          : undefined;
+      const fieldValidateCallback = (firstName: string) =>
+        firstName.toLowerCase().startsWith('foo')
+          ? [
+              {
+                type: 'notFoo',
+                value: firstName,
+                message: 'Foo is an invalid first name!',
+              },
+            ]
+          : undefined;
+
+      await render(<template>
+        <HeadlessForm
+          @data={{data}}
+          @validate={{formValidateCallback}}
+          as |form|
+        >
+          <form.field
+            @name="firstName"
+            @validate={{fieldValidateCallback}}
+            as |field|
+          >
+            <field.label>First Name</field.label>
+            <field.input required pattern="^[A-Za-z]+$" data-test-first-name />
+            <field.errors data-test-first-name-errors as |errors|>
+              {{#each errors as |e index|}}
+                <div data-test-error={{index}}>
+                  <div data-test-error-type>
+                    {{e.type}}
+                  </div>
+                  <div data-test-error-value>
+                    {{e.value}}
+                  </div>
+                  <div data-test-error-message>
+                    {{e.message}}
+                  </div>
+                </div>
+              {{/each}}
+            </field.errors>
+          </form.field>
+          <form.field @name="lastName" as |field|>
+            <field.label>Last Name</field.label>
+            <field.input data-test-last-name />
+            <field.errors data-test-last-name-errors />
+          </form.field>
+          <button type="submit" data-test-submit>Submit</button>
+        </HeadlessForm>
+      </template>);
+
+      await click('[data-test-submit]');
+
+      assert
+        .dom('[data-test-first-name-errors] [data-test-error]')
+        .exists({ count: 3 });
+
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-type]'
+        )
+        .hasText('native');
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-value]'
+        )
+        .hasText('foo123');
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-message]'
+        )
+        .hasAnyText(); // validation error message is browser and locale dependant, so testing against actual message would be very brittle.
```

Comment:
Yeah, there is also btw some kind of DX/UX tradeoff here that we (and UX designers/writers) need to be aware of: while just adding e.g. `required` somewhere is super easy instead of having to pass custom JS validation logic (be they custom functions, or ember-changeset, or yup or whatnot in the future), we do loose the capability to influence the actual wording here!